File: ios/ErrorBook/BookStore.swift
import Foundation
import Combine

// MARK: - 数据库加载/落盘（对应桌面端 src/lib/db.ts + src/store.tsx）
// 数据目录固定为 App 的 Documents/错题本（通过「文件」App / Finder 可见，可与桌面版整目录互通）。

let maxSnapshots = 20
/// 快照至少间隔 1 分钟做一次，避免刷屏
private var lastSnapshotAt: Double = 0

@MainActor
final class BookStore: ObservableObject {
  @Published private(set) var db = Database()
  @Published private(set) var loadError: String?

  let dataDir: URL
  var assetsDir: URL { dataDir.appendingPathComponent("assets") }

  var allTags: [String] { tagCounts.map { $0.0 } }
  var tagCounts: [(String, Int)] {
    var m: [String: Int] = [:]
    for x in db.mistakes { for t in x.tags { m[t, default: 0] += 1 } }
    for t in db.tags where m[t] == nil { m[t] = 0 } // 预建标签计数 0，列表里置灰可点
    return m.sorted { String.naturalLess($0.key, $1.key) }.map { ($0.key, $0.value) }
  }

  init() {
    let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
    dataDir = docs.appendingPathComponent("错题本")
    do {
      try Self.ensureDirs(dataDir)
      db = try Self.loadDb(from: dataDir)
    } catch {
      loadError = "数据目录初始化失败：\(error.localizedDescription)"
    }
  }

  // MARK: 文件层

  nonisolated static func ensureDirs(_ dir: URL) throws {
    let fm = FileManager.default
    for sub in ["assets", "snapshots"] {
      try fm.createDirectory(at: dir.appendingPathComponent(sub), withIntermediateDirectories: true)
    }
  }

  nonisolated static func loadDb(from dir: URL) throws -> Database {
    let path = dir.appendingPathComponent("data.json")
    guard FileManager.default.fileExists(atPath: path.path) else { return Database() }
    do {
      let data = try Data(contentsOf: path)
      return try JSONDecoder().decode(Database.self, from: data)
    } catch {
      NSLog("data.json 解析失败，按空数据处理：\(error)")
      return Database()
    }
  }

  /// 原子写入：先写临时文件再替换；替换前把旧版拷进 snapshots/（限流 + 保留最近 20 份）
  func persist() {
    do {
      try Self.saveDb(db, to: dataDir)
    } catch {
      NSLog("保存失败：\(error)")
    }
  }

  nonisolated static func saveDb(_ db: Database, to dir: URL) throws {
    let fm = FileManager.default
    let dataPath = dir.appendingPathComponent("data.json")
    let now = Date().nowMs
    if fm.fileExists(atPath: dataPath.path), now - lastSnapshotAt > 60_000 {
      lastSnapshotAt = now
      let stamp = ISO8601DateFormatter().string(from: Date()).replacingOccurrences(of: ":", with: "-").replacingOccurrences(of: ".", with: "-")
      let snap = dir.appendingPathComponent("snapshots/data-\(stamp).json")
      try? fm.copyItem(at: dataPath, to: snap)
      pruneSnapshots(dir.appendingPathComponent("snapshots"))
    }
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.prettyPrinted]
    let tmp = dir.appendingPathComponent("data.json.tmp")
    try encoder.encode(db).write(to: tmp, options: .atomic)
    if fm.fileExists(atPath: dataPath.path) {
      _ = try fm.replaceItemAt(dataPath, withItemAt: tmp)
    } else {
      try fm.moveItem(at: tmp, to: dataPath)
    }
  }

  nonisolated private static func pruneSnapshots(_ snapDir: URL) {
    let fm = FileManager.default
    guard let names = try? fm.contentsOfDirectory(atPath: snapDir.path) else { return }
    let sorted = names.filter { $0.hasPrefix("data-") }.sorted()
    guard sorted.count > maxSnapshots else { return }
    for name in sorted.prefix(sorted.count - maxSnapshots) {
      try? fm.removeItem(at: snapDir.appendingPathComponent(name))
    }
  }

  // MARK: 错题 CRUD

  private func mutate(_ f: (inout Database) -> Void) {
    var next = db
    f(&next)
    db = next
    persist()
  }

  func addMistake(_ m: Mistake) {
    mutate { $0.mistakes.append(m) }
  }

  func updateMistake(_ m: Mistake) {
    mutate { d in d.mistakes = d.mistakes.map { $0.id == m.id ? m : $0 } }
  }

  func mistake(id: String) -> Mistake? {
    db.mistakes.first { $0.id == id }
  }

  func deleteMistake(_ id: String) {
    mutate { d in d.mistakes = d.mistakes.filter { $0.id != id } }
  }

  func setMistakeFolders(_ mistakeId: String, _ folderIds: [String]) {
    mutate { d in
      d.mistakes = d.mistakes.map { m in
        m.id == mistakeId ? Mistake(id: m.id, folderIds: folderIds, question: m.question, analysis: m.analysis, tags: m.tags, createdAt: m.createdAt, updatedAt: Date().nowMs) : m
      }
    }
  }

  // MARK: 文件夹 CRUD

  @discardableResult
  func createFolder(name: String, parentId: String?) -> Folder {
    let f = Folder(name: name.trimmingCharacters(in: .whitespacesAndNewlines), parentId: parentId)
    mutate { $0.folders.append(f) }
    return f
  }

  /// 按名称+父级查找，没有就创建（批量导入按源结构落位用）
  @discardableResult
  func findOrCreateFolder(name: String, parentId: String?) -> Folder {
    let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
    if let found = db.folders.first(where: { $0.name == trimmed && $0.parentId ?? nil == parentId ?? nil }) {
      return found
    }
    let f = Folder(name: trimmed, parentId: parentId)
    mutate { $0.folders.append(f) }
    return f
  }

  func renameFolder(_ id: String, _ name: String) {
    mutate { d in
      d.folders = d.folders.map { f in
        f.id == id ? Folder(id: f.id, name: name.trimmingCharacters(in: .whitespacesAndNewlines), parentId: f.parentId, createdAt: f.createdAt) : f
      }
    }
  }

  /// 删除文件夹：从错题的所属列表里移除该文件夹（清空的落到未分类），子文件夹上移一级
  func deleteFolder(_ id: String) {
    mutate { d in
      let parentId = d.folders.first { $0.id == id }?.parentId ?? nil
      d.folders = d.folders.filter { $0.id != id }.map { f in
        f.parentId == id ? Folder(id: f.id, name: f.name, parentId: parentId, createdAt: f.createdAt) : f
      }
      d.mistakes = d.mistakes.map { m in
        m.folderIds.contains(id)
          ? Mistake(id: m.id, folderIds: m.folderIds.filter { $0 != id }, question: m.question, analysis: m.analysis, tags: m.tags, createdAt: m.createdAt, updatedAt: m.updatedAt)
          : m
      }
    }
  }

  // MARK: 笔记 CRUD

  func addNote(_ n: Note) {
    mutate { $0.notes.append(n) }
  }

  func updateNote(_ n: Note) {
    mutate { d in d.notes = d.notes.map { $0.id == n.id ? n : $0 } }
  }

  func deleteNote(_ id: String) {
    mutate { d in d.notes = d.notes.filter { $0.id != id } }
  }

  /// 批量追加错题（合并导入用），跳过已存在的 id，返回实际新增数
  @discardableResult
  func addMistakes(_ ms: [Mistake]) -> Int {
    let ids = Set(db.mistakes.map { $0.id })
    let add = ms.filter { !ids.contains($0.id) }
    if !add.isEmpty { mutate { $0.mistakes.append(contentsOf: add) } }
    return add.count
  }

  @discardableResult
  func addNotes(_ ns: [Note]) -> Int {
    let ids = Set(db.notes.map { $0.id })
    let add = ns.filter { !ids.contains($0.id) }
    if !add.isEmpty { mutate { $0.notes.append(contentsOf: add) } }
    return add.count
  }

  // MARK: 标签

  /// 新建预建标签；空名或已存在（错题已带/已预建）时静默跳过
  func createTag(_ name: String) {
    let t = name.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !t.isEmpty else { return }
    let used = Set(db.mistakes.flatMap { $0.tags })
    guard !used.contains(t), !db.tags.contains(t) else { return }
    mutate { $0.tags.append(t) }
  }

  // MARK: 待导入清单（做题 tab）

  func addPendingImport(_ p: PendingImport) {
    mutate { d in
      guard !d.pendingImports.contains(where: { $0.id == p.id }) else { return }
      d.pendingImports.append(p)
    }
  }

  /// 批量并入（合并导入/拉取用），按 id 去重，返回实际新增数
  @discardableResult
  func addPendingImports(_ ps: [PendingImport]) -> Int {
    let ids = Set(db.pendingImports.map { $0.id })
    let add = ps.filter { !ids.contains($0.id) }
    if !add.isEmpty { mutate { $0.pendingImports.append(contentsOf: add) } }
    return add.count
  }

  func removePendingImport(_ id: String) {
    mutate { d in
      d.pendingImports.removeAll { $0.id == id }
    }
  }

  /// 批量并入预建标签（合并导入用），跳过已有（含错题已带的），返回实际新增数
  @discardableResult
  func addTags(_ names: [String]) -> Int {
    let used = Set(db.mistakes.flatMap { $0.tags }).union(db.tags)
    var seen = Set<String>()
    let add = names
      .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
      .filter { !$0.isEmpty && !used.contains($0) && seen.insert($0).inserted }
    if !add.isEmpty { mutate { $0.tags.append(contentsOf: add) } }
    return add.count
  }
}

File: android/app/src/main/java/com/errorbook/android/BookStore.kt
package com.errorbook.android

import android.content.Context
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.serialization.json.Json
import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

// MARK: - 数据库加载/落盘（对应桌面端 src/lib/db.ts + src/store.tsx、iOS 端 BookStore.swift）
// 数据目录固定为 App 外部私有目录（getExternalFilesDir/错题本）：
// 免存储权限、USB 连电脑可见（MTP），可与桌面版整目录互拷；卸载 App 即删除。

private const val MAX_SNAPSHOTS = 20

class BookStore(context: Context) {

    val db = MutableStateFlow(Database())
    val loadError = MutableStateFlow<String?>(null)
    val dbState: StateFlow<Database> get() = db

    val dataDir: File = File(context.getExternalFilesDir(null) ?: context.filesDir, "错题本")
    val assetsDir: File get() = File(dataDir, "assets")

    /** 写盘串行队列：所有落盘按顺序执行，快速连续 mutate 也不会交错写坏文件 */
    private val ioScope = CoroutineScope(SupervisorJob() + Dispatchers.IO.limitedParallelism(1))
    private val lock = Any()

    val allTags: List<String> get() = tagCounts.map { it.first }

    /** 错题自带标签的计数 + 预建标签（计数 0，列表里置灰可点），按自然排序 */
    val tagCounts: List<Pair<String, Int>>
        get() {
            val m = LinkedHashMap<String, Int>()
            for (x in db.value.mistakes) for (t in x.tags) m[t] = (m[t] ?: 0) + 1
            for (t in db.value.tags) if (!m.containsKey(t)) m[t] = 0
            return m.entries.sortedWith { a, b -> if (naturalLess(a.key, b.key)) -1 else if (naturalLess(b.key, a.key)) 1 else a.key.compareTo(b.key) }
                .map { it.key to it.value }
        }

    init {
        try {
            ensureDirs(dataDir)
            db.value = loadDb(dataDir)
        } catch (e: Exception) {
            loadError.value = "数据目录初始化失败：${e.message}"
        }
    }

    // MARK: 文件层

    fun persist() {
        val snapshot = synchronized(lock) { db.value }
        ioScope.launch {
            try {
                saveDb(snapshot, dataDir)
            } catch (e: Exception) {
                Log.e("BookStore", "保存失败", e)
            }
        }
    }

    /** 改库统一入口：同步换内存态，异步落盘（顺序队列保证最终一致） */
    fun mutate(f: (Database) -> Database) {
        synchronized(lock) { db.value = f(db.value) }
        persist()
    }

    companion object {
        val json = Json {
            prettyPrint = true
            ignoreUnknownKeys = true
            isLenient = true
        }

        fun ensureDirs(dir: File) {
            for (sub in listOf("assets", "snapshots")) File(dir, sub).mkdirs()
        }

        fun loadDb(dir: File): Database {
            val path = File(dir, "data.json")
            if (!path.exists()) return Database()
            return try {
                json.decodeFromString(DatabaseSerializer, path.readText())
            } catch (e: Exception) {
                Log.e("BookStore", "data.json 解析失败，按空数据处理", e)
                Database()
            }
        }

        /** 快照至少间隔 1 分钟做一次，避免刷屏 */
        private var lastSnapshotAt = 0.0

        /** 原子写入：先写临时文件再替换；替换前把旧版拷进 snapshots/（限流 + 保留最近 20 份） */
        fun saveDb(db: Database, dir: File) {
            ensureDirs(dir)
            val dataPath = File(dir, "data.json")
            val now = nowMs()
            if (dataPath.exists() && now - lastSnapshotAt > 60_000) {
                lastSnapshotAt = now
                val fmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
                fmt.timeZone = TimeZone.getTimeZone("UTC")
                val stamp = fmt.format(Date()).replace(":", "-").replace(".", "-")
                try {
                    dataPath.copyTo(File(dir, "snapshots/data-$stamp.json"), overwrite = true)
                } catch (_: Exception) {
                }
                pruneSnapshots(File(dir, "snapshots"))
            }
            val tmp = File(dir, "data.json.tmp")
            tmp.writeText(json.encodeToString(DatabaseSerializer, db))
            if (dataPath.exists()) {
                Files.move(tmp.toPath(), dataPath.toPath(), StandardCopyOption.REPLACE_EXISTING)
            } else {
                Files.move(tmp.toPath(), dataPath.toPath())
            }
        }

        private fun pruneSnapshots(snapDir: File) {
            val sorted = snapDir.listFiles()?.filter { it.name.startsWith("data-") }?.sortedBy { it.name } ?: return
            if (sorted.size <= MAX_SNAPSHOTS) return
            sorted.subList(0, sorted.size - MAX_SNAPSHOTS).forEach { it.delete() }
        }
    }

    // MARK: 错题 CRUD

    fun addMistake(m: Mistake) = mutate { it.copy(mistakes = it.mistakes + m) }

    fun updateMistake(m: Mistake) = mutate { d ->
        d.copy(mistakes = d.mistakes.map { if (it.id == m.id) m else it })
    }

    fun mistake(id: String): Mistake? = db.value.mistakes.firstOrNull { it.id == id }

    fun deleteMistake(id: String) = mutate { d -> d.copy(mistakes = d.mistakes.filter { it.id != id }) }

    fun setMistakeFolders(mistakeId: String, folderIds: List<String>) = mutate { d ->
        d.copy(mistakes = d.mistakes.map { m ->
            if (m.id == mistakeId) m.copy(folderIds = folderIds, updatedAt = nowMs()) else m
        })
    }

    // MARK: 文件夹 CRUD

    fun createFolder(name: String, parentId: String?): Folder {
        val f = Folder(name = name.trim(), parentId = parentId)
        mutate { it.copy(folders = it.folders + f) }
        return f
    }

    /** 按名称+父级查找，没有就创建（批量导入按源结构落位用） */
    fun findOrCreateFolder(name: String, parentId: String?): Folder {
        val trimmed = name.trim()
        db.value.folders.firstOrNull { it.name == trimmed && it.parentId == parentId }?.let { return it }
        val f = Folder(name = trimmed, parentId = parentId)
        mutate { it.copy(folders = it.folders + f) }
        return f
    }

    fun renameFolder(id: String, name: String) = mutate { d ->
        d.copy(folders = d.folders.map { if (it.id == id) it.copy(name = name.trim()) else it })
    }

    /** 删除文件夹：从错题的所属列表里移除该文件夹（清空的落到未分类），子文件夹上移一级 */
    fun deleteFolder(id: String) = mutate { d ->
        val parentId = d.folders.firstOrNull { it.id == id }?.parentId
        d.copy(
            folders = d.folders.filter { it.id != id }.map { if (it.parentId == id) it.copy(parentId = parentId) else it },
            mistakes = d.mistakes.map { m ->
                if (m.folderIds.contains(id)) m.copy(folderIds = m.folderIds.filter { it != id }) else m
            }
        )
    }

    // MARK: 笔记 CRUD

    fun addNote(n: Note) = mutate { it.copy(notes = it.notes + n) }

    fun updateNote(n: Note) = mutate { d -> d.copy(notes = d.notes.map { if (it.id == n.id) n else it }) }

    fun deleteNote(id: String) = mutate { d -> d.copy(notes = d.notes.filter { it.id != id }) }

    // MARK: 批量并入（合并导入/远程拉取用），按 id 去重，返回实际新增数

    fun addMistakes(ms: List<Mistake>): Int {
        var added = 0
        mutate { d ->
            val ids = d.mistakes.map { it.id }.toSet()
            val add = ms.filter { it.id !in ids }
            added = add.size
            if (add.isEmpty()) d else d.copy(mistakes = d.mistakes + add)
        }
        return added
    }

    fun addNotes(ns: List<Note>): Int {
        var added = 0
        mutate { d ->
            val ids = d.notes.map { it.id }.toSet()
            val add = ns.filter { it.id !in ids }
            added = add.size
            if (add.isEmpty()) d else d.copy(notes = d.notes + add)
        }
        return added
    }

    // MARK: 标签

    /** 新建预建标签；空名或已存在（错题已带/已预建）时静默跳过 */
    fun createTag(name: String) {
        val t = name.trim()
        if (t.isEmpty()) return
        val used = db.value.mistakes.flatMap { it.tags }.toSet()
        if (t in used || t in db.value.tags) return
        mutate { it.copy(tags = it.tags + t) }
    }

    // MARK: 待导入清单（做题 tab）

    fun addPendingImport(p: PendingImport) = mutate { d ->
        if (d.pendingImports.any { it.id == p.id }) d else d.copy(pendingImports = d.pendingImports + p)
    }

    fun addPendingImports(ps: List<PendingImport>): Int {
        var added = 0
        mutate { d ->
            val ids = d.pendingImports.map { it.id }.toSet()
            val add = ps.filter { it.id !in ids }
            added = add.size
            if (add.isEmpty()) d else d.copy(pendingImports = d.pendingImports + add)
        }
        return added
    }

    fun removePendingImport(id: String) = mutate { d ->
        d.copy(pendingImports = d.pendingImports.filter { it.id != id })
    }

    /** 批量并入预建标签（合并导入用），跳过已有（含错题已带的），返回实际新增数 */
    fun addTags(names: List<String>): Int {
        var added = 0
        mutate { d ->
            val used = d.mistakes.flatMap { it.tags }.toSet() + d.tags
            val seen = mutableSetOf<String>()
            val add = names.map { it.trim() }.filter { it.isNotEmpty() && it !in used && seen.add(it) }
            added = add.size
            if (add.isEmpty()) d else d.copy(tags = d.tags + add)
        }
        return added
    }
}
